Optimize

satoshi-szk opened the pull request:
- Add reduce function for saving MIDI channels
- clean up codes

satoshi-szk committed:
Clean up code

satoshi-szk committed:
Add reducePartials function

- for saving midi channels

satoshi-szk committed:
Add example of reducePartials

satoshi-szk commented:
Hi @yuichkun, could you review this, too?

yuichkun commented on lib/partials2melodies.js:
```diff
@@ -29,21 +29,21 @@ function partials2melodies(_partials) {
     function detectNoteOnNoteOffIndices(amps) {
-      const result = [];
+      const noteOnOffs = [];
       // 'note on' = 1, 'note off' = -1, 'nothing' = 0
```
This comment is helpful but this still leaves the following mysterious-looking code like `noteOnOffs.push(0)`, which can be just easily rewritten with `noteOnOffs.push(NOTE_OFF)`, if you just predefine constants for them in the local scope.

yuichkun replied:
Javascript doesn't have enums, but you can always at least try to abstract the hard-to-read code with human-readable constants/variables/functions. Be familiar with 説明変数.

yuichkun commented on lib/partials2melodies.js:
```diff
@@ -29,21 +29,21 @@ function partials2melodies(_partials) {
-          else result.push(0);
+          if (amp > 0) noteOnOffs.push(1);
+          else noteOnOffs.push(0);
         } else if (amp > 0 && amps[idx - 1] == 0) {
```
This is a perfect example where you could apply 説明変数.
What does `amp > 0 && amps[idx-1] == 0` represent? 
Be as specific as possible.
Providing explanations doesn't only help the readers understand the logic, but also help yourself organize minds.

yuichkun commented on lib/partials2melodies.js:
```diff
@@ -29,21 +29,21 @@ function partials2melodies(_partials) {
-      const result = [];
+      const noteOnOffs = [];
       // 'note on' = 1, 'note off' = -1, 'nothing' = 0
       amps.forEach((amp, idx, amps) => {
```
I think this whole block can be refactored with `amps.map`, resulting with the elimination of needless `noteOnOffs` variable.

yuichkun replied:
Also if you use `amps.map` your nested if-statements will be much cleaner because you can early-return noteOnOff values, which will also get rid of the last else-statement.

yuichkun commented on lib/partials2melodies.js:
```diff
@@ -29,21 +29,21 @@ function partials2melodies(_partials) {
+      return noteOnOffs;
     }
 
     function calcActiveNoteNums(freqs, noteOnOffs) {
```
Maybe it's just me but,
I think this function name potentially involves two separate confusions.

1. `Nums` could mean both  **sum** or **indices**.
2. `calc` is still ambiguous. Maybe better replaced with something related to the operation.

For example, 
If I were to write a function that collects all the positions of active notes,
I would define `collectActiveNotePos`/`collectActiveNoteIndices`/`findAllActiveNotes`, etc.

If I were to write a function that gets the total number of active notes,
I would define `getSumOfActiveNotes` or something.

yuichkun replied:
However, 
@zigen is more practiced in programming, in general. 
Probably he could come up with more suitable names for these kind of situations.

satoshi-szk committed:
Improve readability with explanatory variable

satoshi-szk committed:
Clean up code by replacing amps.forEach with amps.map

satoshi-szk committed:
Improve readability

- Rename function name
- Rename confusing variable name
- - activeNoteNum to midiNoteNum
- Add explanatory variables
- - prevNoteOnOff
- - closestNoteNum
- - prevMidiNoteNum
- - NOTE_ON, NOTE_OFF, NOTHING

satoshi-szk committed:
Refactor

- Add explanation comment
- Clean up conditions

satoshi-szk commented:
@yuichkun , 
Thank you for your detailed review. Now codes are much cleaner and more readable!!
Can you confirm and approve please?

yuichkun commented:
Looks much cleaner! I love the way you refactored, and also it was really nice that you separated changes into different commits.
